### ISSUE TITLE
Cannot stay logged in after token expires

### ISSUE THREAD

**Issue by WillChilcutt** [okta/okta-sdk-appauth-ios#64]
Hello,

I'm finding problems related to expired tokens.

Let's say I grab a token at 10am. That token expires at 11am (1 hour later) but I can use the refresh token to grab a new access token.

Example situations I've run into:

1. I kill my app at 10:15am and launch it again at 10:30am. In the `OktaAuth.isAuthenticated()` it properly loads the previously created access token as expected and the app works as normal.

2. I leave my app open until 11am, when I check the validity of my token. I see it has expired. I then call `OktaAuth.refresh().then` etc. etc. and as expected I successfully get an access token. Everything works great for my app until I kill it and launch it again. OktaAuth now has a problem retrieving tokens in the `OktaAuth` `isAuthenticated()` function even though I was able to successfully refresh the access token. It fails at unarchiving the Token Manager and thus I cannot refresh the auth token

3. After authenticating at 10am I close my app at 10:15am, with still an access token expiration date of 11am . I launch it at 11:15am, which would mean that my access token expired 15 minutes earlier, so I would expect to be able to use the refresh token to get a fresh access token, but as I found in the previous situation, upon trying unarchive the token manager it fails to set the tokens and thus there is no refresh token to use.

This has basically made our app very hard to use as the `okta-sdk-appauth-ios` has a hard time reloading it's tokens once they have expired on a fresh launch, thus causing each of our users having to log in nearly every time they use the app.

Have the Okta engineers validated that they are able to successfully refresh the app from a cold start with an access token expired?

Thanks

**Comment by jplourde13** [comment]
This also happens to me when using the Android version of this sdk:

https://github.com/okta/okta-sdk-appauth-android

**Comment by samleegray** [comment]
I'm getting the same issue on iOS.

**Comment by jmelberg-okta** [comment]
Thanks for the report @WillChilcutt, @jplourde13, and @samleegray!

This issue seems closely related to #63, where we're seeing consistent issues around unarchiving the object from a saved state.

I'm going to prioritize this issue with our team so it gets addressed quickly.

- Jordan

**Comment by WillChilcutt** [comment]
@jmelberg-okta Is there any ETA on when this fix will be implemented? We are blocked right now as users cannot stay logged in for more than an hour.

**Comment by roger-at-nike** [comment]
Could we get an update on this issue? It's impacting us here at Nike as well.

**Comment by jdberry** [comment]
What's the status on this issue? This is killing us. It's incredible to me that this has seemingly gotten no attention.

**Comment by jmelberg-okta** [comment]
Hi all,

Apologies again for the issues this is causing. We're looking into the root cause of this issue **this week**. Once our analysis is complete, we'll update this issue with a status report.

**Comment by jmelberg-okta** [comment]
Thanks again for your ongoing patience. I was able to identify the root cause of this issue, and plan to release a new version with this fix as soon as I can.

I'm moving tagging this issue as "In Progress" - and will update this issue as work is completed.

**Comment by jdberry** [comment]
@jmaldonado-okta: Are you able to give any summary on the scope of the change and what your ETA is for fixing it? We have a deadline coming up very soon for which we need this fix, and have to evaluate whether to come up with our own solution.

**Comment by jmelberg-okta** [comment]
@jdberry - The fix is pretty straightforward, and you can see the changes in #68.

I was able to validate the changes given @WillChilcutt's reproduction steps above, however I'd still like to spend more time writing tests to cover more flows before this hits `master`.

If you'd like, please pull down the changes and test locally. If anything doesn't act as you'd expect it to, please feel free to tag me here.

Again, thank you all for your patience.

**Comment by jdberry** [comment]
@jmaldonado-okta : that seems to change the behavior a bit, but it still doesn't seem to be working correctly.

We get into this state, for instance, if the app restarts after the access token has timed out. What we do then is the following:

On startup:

 1.  If the OktaAuth.isAuthenticated(), then (we're good)
 2. Otherwise, try to OktaAuth.refresh() before doing anything else (i.e., before falling back to calling login())
  
In this case, in step 2, the refresh appears to succeed, but at that point OktaAuth.isAuthenticated() still says that it's not authenticated. The refresh hasn't stored new tokens in the TokenManager?

Is this the thing we should be doing, or is something else expected?

**Comment by jdberry** [comment]
@jmaldonado-okta : I made a change that seems (on very, very quick inspection) to have fixed the issue I describe above.

I altered the code of OktaRefresh.swift to call storeAuthState(tokens!) just before the final resolve(token) call. I don't know if that's the best fix, but it seems to solve the issue I'm seeing (after applying your patch).

```swift
if error != nil {
    return reject(OktaError.ErrorFetchingFreshTokens(error!.localizedDescription))
}
guard let token = accessToken else {
    return reject(OktaError.ErrorFetchingFreshTokens("Access Token could not be refreshed."))
}

// I added the following line...
OktaAuthorization().storeAuthState(tokens!)

return resolve(token)
```

**Comment by jdberry** [comment]
Incidentally, there's another fix we also had to make. We call Okta using CocoaPods, and it's linked as a framework. Due to this, there's inconsistency in how OktaTokenManager is named, and due to swift's addition of the module name to the class name, the OktaTokenManager cannot sometimes be successfully unarchived from the keychain.

We fixed this problem by doing the following at our app init time:

```swift
        NSKeyedArchiver.setClassName("OktaTokenManager", for: OktaTokenManager.self)
        NSKeyedUnarchiver.setClass(OktaTokenManager.self, forClassName: "OktaTokenManager")
```

Ideally, OktaTokenManager would have an `@objc(OktaTokenManager)` attribute, in order to pin down the name of the class, but note that this could cause existing code/data to not be able to restore at all, so some nuance is needed.

**Comment by WillChilcutt** [comment]
Pulled in both @jmelberg-okta and @jdberry 's changes, which ended up working for us! Thanks!

**Comment by roger-at-nike** [comment]
A mix of the two worked for us as well. Thanks guys! 🙇

**Comment by bynelus** [comment]
Also pulled in both changes, but still won't work. When I login at start it, it succeeds. But when I close the app and remove it from background, it seems the the tokens are not stored resulting in a login every time the app starts. 

Is there anyone who can help out to fix this?

**Comment by jmelberg-okta** [comment]
Hi all,

I want to apologize again for the unacceptably long delay for not getting this resolved sooner. Our team is working on getting a dedicated engineer to maintain this library - something this repository has never had.

This [bug](#68) has now been merged into `master` is is available on Cocoapods as version `1.0.1`.

**Comment by WillChilcutt** [comment]
@jmelberg-okta Is there anything we need to do different on our side as users of this framework with the new `1.0.1` version?

**Comment by jmelberg-okta** [comment]
Hey @WillChilcutt - No changes should be required on your end.

This library was neglecting to re-save the authentication state when refreshing tokens, so if you had any workaround logic for this, it should no longer be required. Therefore, you should be safe to upgrade :)
